### ISSUE TITLE
Readonly properties lost on explicit typing

### ISSUE THREAD

**Issue by niedzielski** [microsoft/TypeScript#41329]
**TypeScript Version:**  4.1.0-dev.20201029

**Search Terms:** readonly properties writable, readonly return type writable

**Code**

Simple example:
```ts
type Card = {suit: string, rank: number}

const readonlyCards: Readonly<Card>[] = [{suit: 'spades', rank: 1}]

// Expected: error.
readonlyCards[0].suit = 'hearts'

// Unexpected: no error. In general, readonly properties are not checked
// when explicit types are specified.
const writableCards: Card[] = readonlyCards

// Unexpected: no error. Card's readonly property attributes were lost.
writableCards[0].suit = 'hearts'
```

More realistic problem:
```ts
// This is not ideal for even trivial examples. Typing at the seams
// (functions and top-level symbols), is usually a safer alternative to
// inferrence since it checks assumptions. For example:
function isHandEmpty(hand: readonly Readonly<Card>[]): boolean {
  return hand[0] // Expected: error. The type checker helped us spot that we
                 // were returning a truthy or falsy value, but not a boolean.
}

// But in the case of readonly properties, seam typing unexpectedly drops the
// readonly attributes. For example:
function shuffle(deck: Readonly<Card>[]): Card[] {
  return deck // Unexpected: readonly on Card properties lost.
}

const shuffledDeck = shuffle(readonlyCards)

// Unexpected: no error.
shuffledDeck[0].suit = 'hearts'
```

Compare to `readonly` arrays:
```ts
// For arrays, at least, the readonly attribute is preserved. For example:
function copy(deck: readonly Card[]): Card[] {
    return deck // Expected: error. The type checker helps us spot that
                // we've broken our commitment to readonly arrays. This will
                // help us avoid errors due to mutating the array copy.
}

copy(shuffledDeck).length = 0
```

**Expected behavior:**
Readonly properties should only be lost when cast away, just like arrays.

**Actual behavior:**
Readonly properties are lost when an explicit type doesn't retain them.

**Playground Link:** https://www.typescriptlang.org/play?ts=4.1.0-dev.20201029#code/C4TwDgpgBAwghgJwCZQLxQN4GcCuBLYALii2ATwDsBzAGigTgoGtiKcBbAIwgQF8AofgGMA9hVL0IcJGIA2IeMizEASlJkV5AHkVIAfAG0AumigHs+IlADkWMNIhZrdBs2IBGXkcEB6H1ABRAA9IIWAIJGIeBBEEADp+BHU5BUQkLAMABiM43AJTawALKQRgJ19-AFUKCBCIMIjWESho2LioAEkKKCoIGoZZF2TNECgwGMhSvEcoRGgKEWAoIWKhJgj+PygAd2Lu2rBZPCF80EgsWaSSULwAM2mkBNFxJe3yYDhOWQhdZVg04ymJLSFK-CpQaoHerhSJQBYtBAxeL-ZC2SQgkZjCY8UCzYBkPCcHDhC7bHjQWQiUgJN4ET7fX5ZHJ5JboIolMrWcEAFUKeAu-LhiygeCQUlkUFusRaADc+lACTK8HAJbU4OxDo52tzwJQqHiFcUSFJ2FhNv4ABS3HAUMJ4MQXRgoYAiMAAWm+colWBAXBEsiwAEo6IKcLgVfJZiQ4LceLNZOEEBQ4MA8HKFSJzSKKLHEX0hNAsJQCyKlit6kxHVhcBrUw72gAxaVqjXfQj8a22uvdfkACSdAVrIAthSdxGBGkjagx2l0hiMgeInBE-qk3Qw-CgkmAOCTUFHFCQTKgW2CoRhUURbSgvOgZ2g5bWceKskgKDD12FwFHrwgm63AGAUBW5bGSVxJDuSZ6lGZDEoUozSrcKo+lAMoqjgEB0ESSwLEscBQMuq6MAkAhZgAQsS2aGg+cBYNAIi3Oik6jOMro4tMWB0HR6oKrq1BQDaUINEgkZIBMFzfn+WwTikeIEthWpQE2CAtEE6qau2nZ2mIJCFDgty3N8FpimsqjDLOaTzouKJHiYG5bhBu7dCZTAnlUNR1MJ47mQh3S6FibFTDMlLUvwpHPBIWB6QZ3xIAAIhWphRfphkQBaMkjL8gbgpCnkXkKCJIgkyUxRECVrEyuSWAUxSIJy4LKZcDAgJxBrfLRwB0JJTGySm8nEtAgrjI4PByo8SnNmprYQJpNrad0ohgMOLneTOoy6MY1kbXZ-4ORAkHOYlp55Y0hXXrevGQMsqzrCpL5gBcH52F+P67cB72gRA1jppwMTrN0Ii7ssIjsOwBDsH0Swuj1mKIM1WDanypJ4LIshve9QFbPdAmOjKIiimdCAXEgGEZlA7DEim0HdXDcCjItIAkYIDMWiVqXxRWgZxN81DfqYmT8EAA

**Related Issues:** 

**Docs:** https://www.typescriptlang.org/docs/handbook/interfaces.html#readonly-properties

**Comment by MartinJohns** [comment]
Duplicate of #13347.

**Comment by niedzielski** [comment]
Thanks, @MartinJohns!
